src/simulators/matrix_product_state/matrix_product_state_internal.cpp

```diff
@@ -1705,13 +1705,39 @@ mps_container_t MPS::move_to_mps_container() {
                                        std::move(q_reg_[i].get_data(1))));
   }
   std::vector<std::vector<double>> lambda_vec;
-  for (auto i=0; i<num_qubits()-1; i++) {
+  for (uint_t i=0; i<num_qubits()-1; i++) {
     ret.second.push_back(std::move(lambda_reg_[i]));
   }
   initialize(MPS());
   return ret;
 }
 
+void MPS::initialize_from_mps(const mps_container_t &mps) {
+
+  uint_t num_qubits = mps.first.size();
+  // clear and restart all internal structures
+  q_reg_.clear();
+  lambda_reg_.clear();
+  q_reg_.resize(num_qubits);
+  lambda_reg_.resize(num_qubits-1);
+  qubit_ordering_.order_.clear();
+  qubit_ordering_.order_.resize(num_qubits);
+  std::iota(qubit_ordering_.order_.begin(), qubit_ordering_.order_.end(), 0);
+
+  qubit_ordering_.location_.clear();
+  qubit_ordering_.location_.resize(num_qubits);
+  std::iota(qubit_ordering_.location_.begin(), qubit_ordering_.location_.end(), 0);
+
+  // initialize values from mps_container_t
+  for (uint_t i=0; i<num_qubits; i++) {
+    MPS_Tensor next_tensor(std::move(mps.first[i].first), std::move(mps.first[i].second));
+    q_reg_[i] = next_tensor;
```

Comment:
Good catch with also moving `mps` internals. The problem is that you cannot move those because `mps` is a `const mps_container_t &`. You would need to get rid of the `const`, assuming it is not used elsewhere. Not sure if that's possible as it comes from a const operation, etc. In addition you would need to create a constructor taking rvalues as params. If it is not possible to pass a `non-const mps_container_t ref`you can remove the `std::move`s in lines 1733 and 1737.

What we can avoid is copying in this line `q_reg_[i] = next_tensor;`
We need to add these constructors accepting rvalues (in matrix_product_state_tensor.hpp) :

```
  MPS_Tensor(MPS_Tensor&& rhs){
    data_ = std::move(rhs.data_);
  }

  MPS_Tensor& operator=(MPS_Tensor&& rhs){
    if (this != &rhs){
      data_ = std::move(rhs.data_);
    }
    return *this;
  }

// Next would only bee needed if we can pass a non-const mps _container_t ref
  MPS_Tensor(cmatrix_t&& data0, cmatrix_t&& data1){ 
    if (!data_.empty())
      data_.clear();
    data_.push_back(data0);
    data_.push_back(data1);
  }
```

And change:
```suggestion
    q_reg_[i] = std::move(next_tensor);
```

Reply:
@vvilpas , thanks for explaining. I was not familiar with the `&&`. A few questions:
1. I still think it is better with the `std::move` on line 1737, since the lambda vectors may also be very big. Why don't we need this?
2. In the code you wrote for the constructors, should we have 
```rhs = nullptr``` after the assignment? as well as for data0, data1.
3. As you said,indeed we cannot change the mps_container_t to be non-const because it is part of the `op` structure. So I didn't include the third cot'r you suggested.

Reply:
No problem!
Regarding questions:
1) The problem is that `mps` is a `const &` so there won't be any actual movement. More or less it's like you mark it as movable by doing `std::move(mps.second[i])`, but as it is `const` the compiler refuse to move it and copies it instead.

2) When an object is moved basically you steal its resources and leave it in some undefined state. In `rhs` we "steal" `data_` which is a `std::vector`, which has its appropriate move constructor to handle the final state of `rhs.data_` (cleared vector), so there is no need to do anything with `rhs.data_`. If `rhs` would have a naked pointer to an array as `data_` member, we could do as you mention: `rhs.data_ = nullptr` so that `rhs` no longer controls `data_`. 
In the case of `data0` and `data1`, they are moved to the vector and the `matrix` move constructor takes care of handling their data pointers, etc.